io-sim/test/Test/Control/Monad/IOSimPOR.hs

```diff
@@ -410,31 +414,64 @@ unit_issue113_nonDep =
 -- Test manually, and supply a small value of n.
 propPermutations :: Int -> Property
 propPermutations n =
-  runST $
   traceNoDuplicates $ \addTrace ->
-  exploreSimTraceST (withScheduleBound 10000) (doit n) $ \_ trace -> do
-    addTrace trace
+  exploreSimTrace (withScheduleBound 10000) (doit n) $ \_ trace ->
+    addTrace trace $
     let Right result = traceResult False trace
-    return $ tabulate "Result" [show $ result] $ True
+     in tabulate "Result" [show result] True
 
 doit :: Int -> IOSim s [Int]
 doit n = do
-  r <- atomically $ newTVar []
+  r <- newTVarIO []
   exploreRaces
   mapM_ (\i -> forkIO $ atomically $ modifyTVar r (++[i])) [1..n]
   threadDelay 1
-  atomically $ readTVar r
+  readTVarIO r
 
 
-traceNoDuplicates :: forall s a prop. (Show a, Testable prop)
-                  => ((a -> ST s ()) -> ST s prop)
-                  -> ST s Property
-traceNoDuplicates k = do
-  v <- newSTRef Map.empty :: ST s (STRef s (Map String Int))
-  prop <- k (\a -> modifySTRef v (Map.insertWith (+) (show a) 1))
-  m <- readSTRef v
-  return $ prop .&&. counterexample (show (Map.keys $ Map.filter (> 1) m)) (maximum m === 1)
+traceNoDuplicates :: (Testable prop1, Show a1) => ((a1 -> a2 -> a2) -> prop1) -> Property
+traceNoDuplicates k = r `pseq` (k addTrace .&&. maximum (traceCounts ()) == 1)
+  where
+    r = unsafePerformIO $ newIORef (Map.empty :: Map String Int)
+    addTrace t x = unsafePerformIO $ do
+      atomicModifyIORef r (\m->(Map.insertWith (+) (show t) 1 m,()))
+      return x
+    traceCounts () = unsafePerformIO $ Map.elems <$> readIORef r
+
+-- | Checks that IOSimPOR is capable of analysing an infinite simulation
+-- lazily.
+--
+-- If the test fails with "outer timeout" error it means that 'exploreSimTrace'
+-- is not lazy and there's something wrong with IOSimPOR implementation and
+-- the trace is being forced way too soon. If, on the other hand, this property
+-- fails with "inner timeout" error, it means that 'exploreSimTrace' is somewhat
+-- lazy, however the trace is not being lazily generated.
+--
+--
+prop_lazy :: Property
+prop_lazy =
+    counterexample "outer timeout"
+  $ within 15000000 -- 15 seconds
+  $ exploreSimTrace
+      id sim $ \_ trace -> do
+        let l = take 10
+              . traceEvents
+              $ trace
```

Comment:
Why not `takeUntil (\(t,_,_,_) -> t < timeLimit)` where `timeLimit` is some random (but positive & small) `Time` generated for this test.

Reply:
That sounds good!